proposals/NNNN-shader-execution-reordering.md

```diff
@@ -0,0 +1,1388 @@
+---------                           | ----------
+`Return: HitObject` | The `HitObject` that contains the result of the initialization operation.
+`RayQuery Query` | RayQuery from which the hit is created.
+
+---
+
+#### HitObject::FromRayQuery with custom attributes
+
+Behaves like the overload of `HitObject::FromRayQuery` that takes only a
+`RayQuery` as argument, with custom attributes associated with
+COMMITTED_PROCEDURAL_PRIMITIVE_HIT. It is ok to always use this overload, even
+for COMMITTED_TRIANGLE_HIT. For anything other than a procedural hit the
+specified attributes are ignored.
+
+```C++
+template<attr_t> 
+static HitObject HitObject::FromRayQuery( 
+    RayQuery Query,
+    attr_t CommittedCustomAttribs);
+```
+
+Parameter                           | Definition
+---------                           | ----------
+`Return: HitObject` | The `HitObject` that contains the result of the initialization operation.
+`RayQuery Query` | RayQuery from which the hit is created.
+`attr_t CommittedCustomAttribs` | See `ReportHit` for definition. If a closesthit shader is invoked from this `HitObject`, `attr_t` must match the attribute type of the closesthit shader. The size of `attr_t` must not exceed `MaxAttributeSizeInBytes` specified in the `D3D12_RAYTRACING_SHADER_CONFIG`.
+
+---
+
+#### HitObject::MakeMiss
+
+Construct a `HitObject` representing a miss, without tracing a ray. It is
+legal to construct a `HitObject` that differs from the result that would be
+obtained by passing the same parameters to `HitObject::TraceRay`, e.g.,
+constructing a miss for a ray that would have hit some geometry if it were
+traced.
+
+```C++
+static HitObject HitObject::MakeMiss(
+    uint RayFlags,
+    uint MissShaderIndex,
+    RayDesc Ray);
+```
+
+Parameter                           | Definition
+---------                           | ----------
+`Return: HitObject` | The `HitObject` that contains the result of the initialization operation.
+`uint RayFlags` | Valid combination of Ray flags as specified by `TraceRay`. Only defined ray flags are propagated by the system.
+`uint MissShaderIndex` | The miss shader index, used to calculate the address of the shader table record. The miss shader index must reference a valid shader table record. Only the bottom 16 bits of this value are used.
+
+---
+
+#### HitObject::MakeNop
+
+Construct a NOP-HitObject that represents neither a hit nor a miss. This is
+the same as a default-initialized `HitObject`. NOP-HitObjects can be useful in
+certain scenarios when combined with `ReorderThread`, e.g., when a thread
+wants to participate in reordering without executing a closesthit or miss
+shader.
+
+```C++
+static HitObject HitObject::MakeNop();
+```
+
+Parameter                           | Definition
+---------                           | ----------
+`Return: HitObject` | The `HitObject` that contains the result of the initialization operation.
+
+---
+
+#### HitObject::Invoke
```

Comment:
Like in comment for `TraceRay`, we are interested in adding more communication about reordering.
We'd consider set of values:

1. PreferNoReorder - (same as for `TraceRay`) the default value with the behavior according to the current spirit. Application doesn't need the reordering happen but the call still is a reorder point and implementation can decide to reorder. Note: This could be also simply spelled out in the spec to be the default if Reorder calls are present.
2. Reorder - with this value, the `[Reorder]HitObject::Invoke( hit, payload )` call would be equivalent to:
   ```c
   ReorderThread( hit ); 
   HitObject::Invoke( hit, payload );
   ```
the value should be limited to shader stages supporting ReorderThread(...).

2. `Reorder(hint,bits)` - calling `[Reorder(hint,bits)]HitObject::Invoke( hit, payload )` would be equivalent to:
   ```c
   ReorderThread( hit, hint, bits); 
   [PreferNoReorder]HitObject::Invoke( hit, payload );
   ```
   Like above the value should be limited to shader stages supporting ReorderThread(...).

We don't propose `NoReorder` here. In case of traceRay, implementation already has to have the inlined raytracing in place so they can probably use it for implementing that. 

The shader invocation depending on implementation may necessary involve reordering as some kind of function call implementation byproduct. 
Such implementation has to deal with difficult choices in compilation. e.g.:
```c++
   ReorderThread(hit, hint, bits);
   color = tex.Sample(sampler, UV);
   HitObject::Invoke( hit, payload );
```
- Is sampler benefiting the reordering before it or not?
- Should I reorder twice (at reorder and at invoke) or nop at `ReorderThread` and reorder according to hint while Invoking the shader?

when separation of `ReorderThread` isn't important part of algorithm, It's easier if application writes it like that:
```c++
   color = tex.Sample(sampler, UV);
   [Reorder(hint, bits)]HitObject::Invoke( hit, payload );
```

Reply:
Again, we do not see a real need for this. If attributes are present at these calls, there is a risk that the user would simply specify [Reorder] on both TraceRay and Invoke, which does not actually mean what we want.

For questions like if Sample benefits from a particular location w.r.t. reordering, the assumption should be that it has an impact and that the application performs the access in the coherence scope where it makes the most sense. However, it does not necessarily hinder fusion. If it is without dependencies, it can still be moved before or after the fused call.

Reply:
Same as above, generally agreed no attributes.